### ISSUE TITLE
Better syntax for OpInfo

### ISSUE THREAD

**Issue by antocuni** [pytorch/pytorch#55159]
This proposal is loosely related to #55152 but I preferred to split them because I think this one is more controversial.

The `op_db` database is getting bigger and bigger and it's becoming hard to navigate inside `common_method_invocations.py`. In particular, one sub-optimal aspect is that the definition of the `sample_inputs` functions are far away from the actual `OpInfo` call. One possible solution is to completely change the syntax, e.g. something along these lines:

```python
class flip(OpInfo):
    op = torch.flip
    dtypes = all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16)
    supports_out = False

    def sample_inputs(self):
        # this can also be a generator, might be easier to yield multiple 
        # samples instead of manually accumulating them into a list
        return [...]
```

The advantage of keeping `sample_inputs` close to the rest of `OpInfo` is obvious, and it will become even more important if in the future we will ever need to introduce other functions like that. E.g. we already have `ref` which often requires a `lambda`.

Moreover, a class-based opdb is more flexible. Consider the following existing code, which contains a lot of duplication:
```python
    OpInfo('div',
           variant_test_name='true_rounding',
           dtypes=all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16),
           sample_inputs_func=partial(sample_inputs_div, rounding_mode='true'),
           skips=(SkipInfo('TestOpInfo', 'test_duplicate_method_tests'),),
           assert_autodiffed=True),
    OpInfo('div',
           variant_test_name='trunc_rounding',
           dtypes=all_types_and(torch.half, torch.bfloat16),
           sample_inputs_func=partial(sample_inputs_div, rounding_mode='trunc'),
           skips=(SkipInfo('TestOpInfo', 'test_duplicate_method_tests'),),
           assert_autodiffed=True),
    OpInfo('div',
           variant_test_name='floor_rounding',
           dtypes=all_types_and(torch.half, torch.bfloat16),
           sample_inputs_func=partial(sample_inputs_div, rounding_mode='floor'),
           skips=(SkipInfo('TestOpInfo', 'test_duplicate_method_tests'),),
           assert_autodiffed=True),
```

With class-based syntax it might become something along these lines:
```python
class div_base(OpInfo):
    op_name = 'div'
    rounding_mode = None
    skips=(SkipInfo('TestOpInfo', 'test_duplicate_method_tests'),),
    assert_autodiffed=True
    def sample_inputs(self):
        yield SampleInput(..., rounding_mode=self.rounding_mode)

class div_true_rounding(div_base):
    rounding_mode = 'true'
    dtypes = all_types_and_complex_and(torch.bool, torch.half, torch.bfloat16)

class div_trunc_rounding(div_base):
    rounding_mode = 'trunc'
    dtypes=all_types_and(torch.half, torch.bfloat16)

class div_floor_rounding(div_base):
    rounding_mode = 'floow'
    dtypes=all_types_and(torch.half, torch.bfloat16)
```

An open question is how to "register" the OpInfo subclasses inside the db. There are at least two options:
1. use a metaclass, e.g. something along these lines (but engineered better :)):
```python
class OpInfoMeta(type):
    def __new__(self, name, bases, dic):
        ...
        op_db.append(self())

class OpInfo(metaclass=OpInfoMeta):
    ...
```

2. use some kind of class decorator:
```python
@op_db.register
class div(OpInfo):
    ...
```

Solution 2 is probably my preferred because it's less magical and it makes it trivial to define multiple `op_db` databases (and/or attach the same `OpInfo` to multiple databases). It still probably requires a small bit of magic to ensure that we don't forget the `@some_db.register` decorator but this should be easy.

Finally, another advantage of this scheme is that it makes it easier to split the `OpInfo` in multiple files, especially considering that `common_method_invocations.py` is almost 5000 lines by now. For example, I am thinking of something like this:

> #### `common_methods_invocation.py`
``` python
...
op_db = OpDb()
import .opinfo_torch # side effects
import .opinfo_torch_linalg # side effects
```

> #### `op_info_torch.py`
``` python
...
@op_db.register
class add(OpInfo):
    ...

@op_db.register
class div(OpInfo)
    ...
```

> #### `op_info_torch_linalg.py`
``` python
...
@op_db.register
class linalg_det(OpInfo):
    ...

@op_db.register
class linalg_svd(OpInfo)
    ...
```

If there is interest in such a thing, I'd be happy to work on it.

/cc @mruberry

**Comment by Lezcano** [comment]
I really like this!
On allowing generators for the samples, this is related: https://github.com/pytorch/pytorch/pull/53044

**Comment by mruberry** [comment]
These are two interesting ideas:

- let's write OpInfos using classes to consolidate code
- let's split up common_methods_invocations.py because it's becoming too large

Let's talk about the second idea first.

Refactoring common_methods_invocations.py sounds like a good PR or series of PRs. We should be careful to pick a reasonable naming scheme for the new files. For example, "unary_ufuncs_ops.py," which contains OpInfos that correspond with "test_unary_ufuncs.py". Similarly "linalg_ops.py" would correspond with "test_linalg.py". Working with generated tests can be confusing, and engineers often struggle with how all these generated files are connected, but I think an approach like that would be clear and scalable.

As far as rewriting the OpInfos as custom classes I'm not in love with the idea. What if we just continue to define OpInfos like we do today but they add themselves to the op_db list? That would still allow their sample inputs to be defined next to them. Creating a class to encapsulate the metadata and sample inputs func seems basically the same except we'd have to instantiate all the classes and add the instantiations to the list, which might be an unnecessary step? I guess I'd like to see what the refactor + auto-registration looks like first before we consider pursuing custom classes.

What are your thoughts, @antocuni?

**Comment by antocuni** [comment]
> Refactoring common_methods_invocations.py sounds like a good PR or series of PRs. We should be careful to pick a reasonable naming scheme for the new files. For example, "unary_ufuncs_ops.py," which contains OpInfos that correspond with "test_unary_ufuncs.py". Similarly "linalg_ops.py" would correspond with "test_linalg.py". Working with generated tests can be confusing, and engineers often struggle with how all these generated files are connected, but I think an approach like that would be clear and scalable.

yes, I agree that the naming scheme is important and it should match the naming of the tests. There are cases in which it's not 100% possible thoughts. E.g. currently `test_ops.py` runs against *all* ops, so there is no way to split it unless we want to unnecessarily duplicate code.

> As far as rewriting the OpInfos as custom classes I'm not in love with the idea. What if we just continue to define OpInfos like we do today but they add themselves to the op_db list (maybe dunder init on the OpInfo base could add them)? That would still allow their sample inputs to be defined next to them. 

I think this would work and it's an improvement over the current situation. However, I fear that implicitly registering the op inside `OpInfo.__init__` is a bit unexpected for the casual user, and there is no hint that this is happening unless you go reading the `__init__` code. By contrast, something like `@op_db.register` makes it clearer that something is happening under the hood, and it's easy to find the relevant code if you want to see what's happening.

> Creating a class to encapsulate the metadata and sample inputs func seems basically the same except we'd have to instantiate all the classes and add the instantiations to the list, which might be an unnecessary step? I guess I'd like to see what the refactor + auto-registration looks like first before we consider pursuing custom classes.

I think that all considered both solutions are very similar and at the end of the day it's just a matter of personal preference over this or that syntax, so it's up to you I suppose :)
Personally, I still prefer the class-based approach because it automatically creates a namespace where to put all the functions/helpers related to a single op, instead of having to write things like `linalg_det_sample_inputs` in the global namespace.

**Comment by mruberry** [comment]
Summary of offline conversation (please correct me if I get this wrong, @antocuni):

- refactoring common_methods_invocations and providing a way for OpInfos to register themselves globally (and explicitly) would be great
- let's do that first and then we can reconsider whether to write them as classes or not
